docker/src/main/java/clocker/docker/location/DockerLocation.java

```diff
@@ -245,13 +239,20 @@ public MachineLocation obtain(Map<?,?> flags) throws NoMachinesAvailableExceptio
         Entities.waitForServiceUp(dockerHost);
 
         // Obtain a new Docker container location, save and return it
-        if (LOG.isDebugEnabled()) {
+        try {
             LOG.debug("Obtain a new container from {} for {}", machine, entity);
+            Map<?,?> hostFlags = MutableMap.copyOf(flags);
+            DockerContainerLocation container = machine.obtain(hostFlags);
+            containers.put(machine, container.getId());
+            return container;
+        } finally {
+            // Release any placement strategy locks
+            for (DockerAwarePlacementStrategy strategy : Iterables.concat(strategies, entityStrategies)) {
+                if (strategy instanceof WithMutexes) {
+                    ((WithMutexes) strategy).releaseMutex(entity.getApplicationId());
```

Comment:
Yes, probably needs more testing and thought given to non-happy-path cases. The lock entry has a timeout to prevent blocking forever, but still not optimal.